src/main/java/org/spongepowered/common/network/channel/TransactionStore.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.network.channel;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.api.network.EngineConnection;
+import org.spongepowered.api.network.channel.TimeoutException;
+
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+/**
+ * A storage for transaction id mappings.
+ */
+public final class TransactionStore {
+
+    private final Supplier<EngineConnection> connection;
+
+    private final AtomicInteger counter = new AtomicInteger();
+    private final ConcurrentMap<Integer, Entry> lookup = Caffeine.newBuilder()
+            .expireAfterAccess(1, TimeUnit.MINUTES)
+            .removalListener((RemovalListener<Integer, Entry>) (key, value, cause) -> {
+                if (cause == RemovalCause.EXPIRED && value != null) {
+                    value.getChannel().handleTransactionResponse(
+                            getConnection(), value.getData(), TransactionResult.failure(new TimeoutException()));
+                }
+            })
+            .build().asMap();
```

Comment:
Does `asMap` render the cache as invalid? Haven't read the javadoc on caffeine in a moment, but I'd figure it would make a copy as a Map.

Reply:
`asMap` returns a view of the cache